server/src/main/java/org/opensearch/search/aggregations/DefaultAggregationProcessor.java

```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public void preProcess(SearchContext context) {
+        try {
+            if (context.aggregations() != null) {
+                if (context.aggregations().factories().hasNonGlobalAggregator()) {
+                    context.queryCollectorManagers()
+                        .put(NonGlobalAggCollectorManager.class, new NonGlobalAggCollectorManagerWithSingleCollector(context));
+                }
+                // initialize global aggregators as well, such that any failure to initialize can be caught before executing the request
+                if (context.aggregations().factories().hasGlobalAggregator()) {
+                    context.queryCollectorManagers()
+                        .put(GlobalAggCollectorManager.class, new GlobalAggCollectorManagerWithSingleCollector(context));
+                }
+            }
+        } catch (IOException ex) {
+            throw new AggregationInitializationException("Could not initialize aggregators", ex);
+        }
+    }
+
+    @Override
+    public void postProcess(SearchContext context) {
+        if (context.aggregations() == null) {
+            context.queryResult().aggregations(null);
+            return;
+        }
+
+        if (context.queryResult().hasAggs()) {
+            // no need to compute the aggs twice, they should be computed on a per context basis
+            return;
+        }
+
+        final AggregationCollectorManager nonGlobalCollectorManager = (AggregationCollectorManager) context.queryCollectorManagers()
+            .get(NonGlobalAggCollectorManager.class);
+        final AggregationCollectorManager globalCollectorManager = (AggregationCollectorManager) context.queryCollectorManagers()
+            .get(GlobalAggCollectorManager.class);
+        try {
+            if (nonGlobalCollectorManager != null) {
+                nonGlobalCollectorManager.reduce(List.of()).reduce(context.queryResult());
+            }
+
+            try {
+                if (globalCollectorManager != null) {
+                    Query query = context.buildFilteredQuery(Queries.newMatchAllQuery());
+                    if (context.getProfilers() != null) {
+                        context.getProfilers()
+                            .addQueryProfiler()
+                            .setCollector((InternalProfileComponent) globalCollectorManager.newCollector());
+                    }
+                    context.searcher().search(query, globalCollectorManager.newCollector());
+                    globalCollectorManager.reduce(List.of()).reduce(context.queryResult());
+                }
+            } catch (Exception e) {
+                throw new QueryPhaseExecutionException(context.shardTarget(), "Failed to execute global aggregators", e);
+            }
+            // create the final result with pipeline tree for versions 2.x which needs to be done after reduce of both global/non-global
+            // aggs
+            finalizeAggregationResults(context);
```

Comment:
This is the new change needed for 2.x

Reply:
Uh ... that was unexpected (good comment), but why we cannot do that in `AggregationReduceableSearchResult`?

Reply:
Updated it to handle in `AggregationReduceableSearchResult`. Earlier for some reason I thought it will build the `pipelineTree` compared to other way where only supplier needs to be provided. Thanks for the comment!